velox/experimental/cudf/exec/VeloxCudfInterop.cpp

```diff
@@ -0,0 +1,184 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/experimental/cudf/exec/NvtxHelper.h"
+#include "velox/experimental/cudf/exec/Utilities.h"
+#include "velox/experimental/cudf/exec/VeloxCudfInterop.h"
+
+#include "velox/common/memory/Memory.h"
+#include "velox/type/Type.h"
+#include "velox/vector/BaseVector.h"
+#include "velox/vector/ComplexVector.h"
+#include "velox/vector/DictionaryVector.h"
+#include "velox/vector/FlatVector.h"
+#include "velox/vector/arrow/Bridge.h"
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/interop.hpp>
+#include <cudf/strings/string_view.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+#include <nvtx3/nvtx3.hpp>
+#include <thrust/copy.h>
+#include <thrust/execution_policy.h>
+
+#include <arrow/c/bridge.h>
+#include <arrow/io/interfaces.h>
+#include <arrow/table.h>
+
+#include <functional>
+#include <numeric>
+
+namespace facebook::velox::cudf_velox {
+
+namespace with_arrow {
+
+std::unique_ptr<cudf::table> toCudfTable(
+    const facebook::velox::RowVectorPtr& veloxTable,
+    facebook::velox::memory::MemoryPool* pool,
+    rmm::cuda_stream_view stream) {
+  // Need to flattenDictionary and flattenConstant, otherwise we observe issues
+  // in the null mask.
+  ArrowOptions arrowOptions{true, true};
+  ArrowArray arrowArray;
+  exportToArrow(
+      std::dynamic_pointer_cast<facebook::velox::BaseVector>(veloxTable),
+      arrowArray,
+      pool,
+      arrowOptions);
+  ArrowSchema arrowSchema;
+  exportToArrow(
+      std::dynamic_pointer_cast<facebook::velox::BaseVector>(veloxTable),
+      arrowSchema,
+      arrowOptions);
+  auto tbl = cudf::from_arrow(&arrowSchema, &arrowArray, stream);
+
+  // Release Arrow resources
+  if (arrowArray.release) {
+    arrowArray.release(&arrowArray);
+  }
+  if (arrowSchema.release) {
+    arrowSchema.release(&arrowSchema);
+  }
+  return tbl;
+}
+
+namespace {
+
+void toSignedIntFormat(char* format) {
+  VELOX_CHECK_NOT_NULL(format);
+  switch (format[0]) {
+    case 'C':
+      format[0] = 'c';
+      break;
+    case 'S':
+      format[0] = 's';
+      break;
+    case 'I':
+      format[0] = 'i';
+      break;
+    case 'L':
+      format[0] = 'l';
+      break;
+    default:
+      return;
+  }
+  printf(
+      "Warning: arrowSchema.format: %s, unsigned is treated as signed indices\n",
+      format);
+}
+
+// Changes all unsigned indices to signed indices for dictionary columns from
+// cudf which uses unsigned indices, but velox uses signed indices.
+void fixDictionaryIndices(ArrowSchema& arrowSchema) {
```

Comment:
We should have unit test to cover each path, looks like this is not tested. And in Velox, every public function like `toVeloxColumn` should have its own test, but I'm not sure if we have the strict requirement for cudf extension.

Reply:
removed the fixDictionaryIndices because cudf changed indices of dictionary to signed.